### PR TITLE
Migrate Portal React component to Typescript

### PR DESCRIPTION
**Pull request by lb-**
- medium change to the `Portal` component but means this is now in Typescript so we get some type safety
- adds react-dom types so that Portal is correctly recognised
- Split from https://github.com/wagtail/wagtail/pull/8618

### CODE REVIEW

File: client/src/components/Portal/Portal.tsx [hunk 1]
```diff
@@ -16,10 +42,11 @@ class Portal extends Component {
     this.onCloseEvent = this.onCloseEvent.bind(this);
   }
 
-  onCloseEvent(e) {
+  onCloseEvent(event: Event) {
```
Comment (kaedroho) [inline client/src/components/Portal/Portal.tsx:45]: I think this should be ``MouseEvent``

**Review by kaedroho — CHANGES_REQUESTED** [review]
Just a minor tweak to an event name, but other than that, this looks good to me!

I have a couple of general React questions:

 - Should we get all our components using the new function style? (This should remove the need for ``defaultProps`` in this case)
 - I'm assuming we are keeping ``propTypes`` in case any untyped JS code uses this component, once we've converted everything to typescript, should we remove ``propTypes``?

File: client/src/components/Portal/Portal.tsx (hunk 1 above)
Comment (lb-) [inline client/src/components/Portal/Portal.tsx:45]:
```suggestion
  onCloseEvent(event: MouseEvent) {
```

**Comment by lb-** [comment]
>     * Should we get all our components using the new function style? (This should remove the need for `defaultProps` in this case)
I think it depends on the component's requirements. Function components are great but once you start using hooks it becomes a bit of a minefield with more edge cases to consider, harder (ish) to unit test and also must enforce the rules of hooks (which is currently disabled in our linting). 
Maybe our team at my company have been burnes too much but we use hooks sparingly and instead pull stuff up into a nearby class component where appropriate. 

>     * I'm assuming we are keeping `propTypes` in case any untyped JS code uses this component, once we've converted everything to typescript, should we remove `propTypes`?
Good question - I went for the least controversial route to Typescript here and did not think too much about typing with proptypes vs TS types. 
I'm happy to remove proptypes and go all in with just typescript here though - probably less duplication. 

As for compatibility - proptypes for JS only use is not really a goal as we only expose the production build of React to those that use Wagtail and that strips out all proptypes and any warnings.

**Comment by lb-** [comment]
After reading the code again I think we should keep propTypes in use for now as the only usage is within Draftail and all those components use JS not TS.

Longer term - we may not even need the Portal component if we do adopt tippy.js in all locations (which is used in some other React components)

**Review by kaedroho — APPROVED** [review]
This looks good to me, thanks!

**Comment by lb-** [comment]
Thanks @kaedroho
